src/SSHDebugPS/PipeConnection.cs

```diff
@@ -0,0 +1,251 @@
+            }
+
+            _shellExecutionManager = new ShellExecutionManager(CreateShellFromSettings(_settings, _outerConnection));
+        }
+
+        public override void BeginExecuteAsyncCommand(string commandText, bool runInShell, IDebugUnixShellCommandCallback callback, out IDebugUnixShellAsyncCommand asyncCommand)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override int ExecuteCommand(string commandText, int timeout, out string commandOutput)
+        {
+            if (_isClosed)
+            {
+                throw new ObjectDisposedException(nameof(PipeConnection));
+            }
+
+            return _shellExecutionManager.ExecuteCommand(commandText, timeout, out commandOutput);
+        }
+
+        public override void Close()
+        {
+            _isClosed = true;
+
+            lock (_lock)
+            {
+                foreach (IRawShell rawShell in _shellList)
+                {
+                    rawShell.Dispose();
+                }
+
+                _shellList.Clear();
+            }
+        }
+
+        protected IRawShell CreateShellFromSettings(IPipeTransportSettings settings, Connection outerConnection, bool isCommandShell = false)
+        {
+            IRawShell rawShell;
+            if (_outerConnection == null)
+            {
+                if (isCommandShell)
+                    rawShell = new CommandShell(settings.ExeCommand, settings.ExeCommandArgs);
+                else
+                    rawShell = new RawLocalShell(settings.ExeCommand, settings.ExeCommandArgs);
+            }
+            else
+            {
+                rawShell = new RawRemoteShell(settings.ExeCommand, settings.ExeCommandArgs, outerConnection);
+            }
+
+            LinkedListNode<IRawShell> node;
+            lock (_lock)
+            {
+                node = _shellList.AddLast(rawShell);
+            }
+
+            rawShell.Closed += (sender, eventArgs) =>
+            {
+                if (_isClosed)
+                    return;
+
+                lock (_lock)
+                {
+                    if (_isClosed)
+                        return;
+
+                    if (node != null)
+                    {
+                        _shellList.Remove(node);
+                        node = null;
+                    }
+                }
+            };
+
+            return rawShell;
+        }
+
+        public override void ExecuteSyncCommand(string commandDescription, string commandText, out string commandOutput, int timeout, out int exitCode)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override string MakeDirectory(string path)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(path), nameof(path));
+
+            if (!DoesCommandExist("mkdir"))
+            {
+                throw new CommandFailedException("mkdir");
+            }
```

Comment:
I think this `DoesCommandExist` pattern is a bad idea in general.  Just try to run the command and handle "doesn't exist" like any other failure: by presenting the error text to the user.